Tasks/Maven/maventask.ts

```diff
@@ -303,70 +333,110 @@ function cleanDirectory(targetDirectory:string):Q.Promise<void> {
         tl.debug('Staging ' + fileToUpload + ' to ' + stagingFilePath);
         // Execute the copy operation. -f overwrites if there is already a file at the destination.
         tl.cp('-f', fileToUpload, stagingFilePath);
     });
 
-    console.log('Uploading artifacts for ' + analysisResult.toolName + ' from ' + toolStagingDir);
+    console.log('Uploading artifacts for ' + toolName + ' from ' + localStagingDir);
 
     // Begin artifact upload - this is an asynchronous operation and will finish in the future
     tl.command("artifact.upload", {
-        containerfolder: 'codeAnalysis',
+        // Put the artifacts in subdirectories corresponding to their module
+        containerfolder: moduleAnalysis.moduleName,
         // Prefix the build number onto the upload for convenience
-        artifactname: buildNumber + '_' + analysisResult.toolName
-    }, toolStagingDir);
+        artifactname: buildNumber + '_' + moduleAnalysis.moduleName + '_' + toolName
+    }, localStagingDir);
 }
 
 // Extract data from code analysis output files and upload results to build server
 function uploadCodeAnalysisResults():void {
     // Need to run this if any one of the code analysis tools were run
-    if (isCodeAnalysisUploadRequired()) {
-        var analysisResults:ar.AnalysisResult[] = [];
-
-        // PMD
-        if (tl.getInput('pmdAnalysisEnabled', true) == 'true') {
-            try {
-                var pmdResults:ar.AnalysisResult = pmd.processPmdOutput(sourcesDir);
-                if (pmdResults) {
-                    analysisResults.push(pmdResults);
+    var enabledCodeAnalysisTools = getEnabledCodeAnalysisTools();
+    if (enabledCodeAnalysisTools.length > 0) {
+        // Discover maven modules
+        var modules:ma.ModuleAnalysis[] = findModules(sourcesDir);
+
+        // Special case: if the root turns up as a module, the automatic name won't do
+        modules.forEach((module) => {
+            if (module.rootFolder == sourcesDir) {
+                module.moduleName = 'root'
+            }
+        });
+
+        tl.debug('Discovered ' + modules.length + ' Maven modules to upload results from.');
+
+        // Discover output files, summarise per-module results
+        modules.forEach((module) => {
+            // PMD
+            if (enabledCodeAnalysisTools.indexOf(pmd.toolName) > -1) {
+                try {
+                    var pmdResults:ar.AnalysisResult = pmd.processPmdOutput(module.rootFolder);
+                    if (pmdResults) {
+                        module.analysisResults[pmdResults.toolName] = pmdResults;
+                    }
+                } catch(e) {
+                    tl.error(e.message);
+                    tl.exit(1);
                 }
-            } catch(e) {
-                tl.error(e.message);
-                tl.exit(1);
             }
-        }
 
-        // Process analysis results - upload files and generate build summary lines
+        });
```

Comment:
Each module has its own, independent /target/ folder, so each moduleAnalysis object is processed independently. The resulting AnalysisResult objects are stored in the analysisResults dictionary field of the moduleAnalysis object, associated by the toolname.